File: app/src/main/java/com/app/id/controller/BaseController.java
package com.app.id.controller;

import com.app.id.application.Application;
import com.app.id.exception.NetworkConnectivityException;

import rx.Observable;
import rx.Single;
import rx.functions.Func0;

class BaseController {

    protected Application app;
    protected PreferencesManager preferencesManager;

    public BaseController(Application app, PreferencesManager preferencesManager) {
        this.app = app;
        this.preferencesManager = preferencesManager;
    }

    protected <T> Observable<T> checkConnectivity(final Observable<T> observable) {
        return Observable.defer(new Func0<Observable<T>>() {
            @Override
            public Observable<T> call() {
                if (app == null || app.isNetworkConnected()) {
                    return observable;
                }

                return Observable.error(NetworkConnectivityException.INSTANCE);
            }
        });
    }

    protected <T> Single<T> checkConnectivity(final Single<T> observable) {
        return Single.defer(new Func0<Single<T>>() {
            @Override
            public Single<T> call() {
                if (app == null || app.isNetworkConnected()) {
                    return observable;
                }

                return Single.error(NetworkConnectivityException.INSTANCE);
            }
        });
    }
}

File: app/src/main/java/com/app/id/controller/BaseController.kt
package com.app.id.controller

import com.app.id.application.Application
import com.app.id.exception.NetworkConnectivityException
import rx.Observable
import rx.Single

open class BaseController(protected var app: Application?, protected var preferencesManager: PreferencesManager) {

    protected fun <T> checkConnectivity(observable: Observable<T>): Observable<T> {
        return Observable.defer {
            if (app == null || app!!.isNetworkConnected) {
                observable
            } else {
                Observable.error<T>(NetworkConnectivityException.INSTANCE)
            }
        }
    }

    protected fun <T> checkConnectivity(observable: Single<T>): Single<T> {
        return Single.defer {
            if (app == null || app!!.isNetworkConnected) {
                observable
            } else {
                Single.error<T>(NetworkConnectivityException.INSTANCE)
            }
        }
    }
}
